[lldb] Add SB API to make a breakpoint a hardware breakpoint

JDevlieghere opened the pull request:
This adds SBBreakpoint::SetIsHardware, allowing clients to mark an existing breakpoint as a hardware breakpoint purely through the API. This is safe to do after creation, as the hardware/software distinction doesn't affect how breakpoint locations are selected.

In some cases (e.g. when writing a trap instruction would alter program behavior), it's important to use hardware breakpoints. Ideally, we’d extend the various `Create` methods to support this, but given their number, this patch limits the scope to the post-creation API. As a workaround, users can also rely on target.require-hardware-breakpoint or use the `breakpoint set` command.

rdar://153528045

jasonmolenda commented:
just a quick sniff test I tried

```
(lldb) process plugin packet send Z0,100000468,4
  packet: Z0,100000468,4
response: OK
(lldb) process plugin packet send z1,100000468,4
  packet: z1,100000468,4
response: OK

a.out`main:
->  0x100000464 <+4>:  sub    sp, sp, #0x20
    0x100000468 <+8>:  stp    x29, x30, [sp, #0x10]
    0x10000046c <+12>: add    x29, sp, #0x10
(lldb) c
Process 73263 resuming
Hello World!
Process 73263 exited with status = 0 (0x00000000) 
(lldb) 
``` 

so maybe this is OK.

JDevlieghere committed:
[lldb] Add SB API to make a breakpoint a hardware breakpoint

This adds SBBreakpoint::SetIsHardware, allowing clients to mark an
existing breakpoint as a hardware breakpoint purely through the API.
This is safe to do after creation, as the hardware/software distinction
doesn't affect how breakpoint locations are selected.

In some cases (e.g. when writing a trap instruction would alter program
behavior), it's important to use hardware breakpoints. Ideally, we’d
extend the various Create methods to support this, but given their
number, this patch limits the scope to the post-creation API. As a
workaround, users can also rely on target.require-hardware-breakpoint or
use the breakpoint set command.

rdar://153528045

jimingham commented on lldb/include/lldb/API/SBBreakpoint.h:
```diff
@@ -148,6 +148,8 @@ class LLDB_API SBBreakpoint {
 
   bool IsHardware() const;
 
+  bool SetIsHardware(bool is_hardware);
```
Document the return value's meaning somewhere.

jimingham commented on lldb/include/lldb/Breakpoint/BreakpointLocation.h:
```diff
@@ -69,7 +69,7 @@ class BreakpointLocation
 
   /// If \a enabled is \b true, enable the breakpoint, if \b false disable it.
-  void SetEnabled(bool enabled);
+  bool SetEnabled(bool enabled);
```
Might be good here to document how this can fail.  If you enable a HW breakpoint after you've used up all the resources with other breakpoints is the only case I can think of, but that's not entirely obvious.  That's the only way this can fail that I can think of.  If there are other ways this could fail, we should return an SBError to distinguish between them.

jimingham commented:
It would be good to test the response to enabling a HW breakpoint or trying to set it to HW when you've exhausted the number of HW breakpoint resources.  For some reason we have SBProcess::GetNumSupportedHardwareWatchpoints but not SBProcess::GetNumSupportedHardwareBreakpoints.  But you could just keep setting HW breakpoints till you get an error.

jimingham commented:
There's one more tricky bit we need to handle here.  What happens if I have a breakpoint that has more locations than we have hardware resources, and you do `SBBreakpoint.SetIsHardware(True)`?  Do we set the first `n` breakpoints and stop there?  Do we back out and restore the breakpoint to software?

Maybe we need an actual error and not a bool to explain what went on there?

JDevlieghere commented:
Please take another look. I improved the error handling and updated the test to test both the case where it fails because HW breakpoint are not available and when it succeeds and they are.

labath commented:
> For some reason we have SBProcess::GetNumSupportedHardwareWatchpoints but not SBProcess::GetNumSupportedHardwareBreakpoints.

Just noting that the values returned from these (partially imaginary) APIs would be hard to use correctly. The two resources might be inter-dependent. E.g. intel can have four hardware breakpoints OR four hardware watchpoints. However, it cannot have four HW breakpoints AND four HW watchpoints. And it might not be able to set even all four as the (linux) kernel might be using some of the registers for its own purposes (something to do with perf, I believe).

jimingham commented on lldb/source/Breakpoint/Breakpoint.cpp:
```diff
@@ -251,6 +251,44 @@ const lldb::TargetSP Breakpoint::GetTargetSP() {
+  }
+
+  if (num_failures != 0)
+    return llvm::createStringError(
```
The locations we didn't convert to hardware will be left disabled.  That's a useful but not obvious bit of information, which it would be good to include in the error string.

Other than that this looks okay to me.

jimingham reviewed (approved):
LGTM.
